resource/field_extractor.go

```diff
@@ -0,0 +1,278 @@
+// Copyright © 2016 Asteris, LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resource
+
+import (
+	"reflect"
+
+	"github.com/pkg/errors"
+)
+
+// FieldMap represents a map of field names to interfaces
+type FieldMap map[string]interface{}
+
+/* Rules for field access in structures:
+
+- Fields that are tagged with `export` will be exported.
+
+- Named structs that are tagged with `export` will be exported as a struct
+
+- Embedded structs will have their exported fields exported in the namespace of
+	the containing struct
+
+- Embedded interfaces will not be exported, nor have their fields exported
+
+- If an embedded struct field name collides with a field from the struct that
+	it's embedded in, both will be exported with the embedded struct being
+	accessible with 'StructName.FieldName' */
+
+var (
+	// ErrNilStruct is returned if there is an attempt to introspect a nil value
+	ErrNilStruct = errors.New("cannot inspect types with a nil value")
+
+	// ErrNonStructIntrospection is returned when attempting to introspect fields
+	// on a non-struct type.
+	ErrNonStructIntrospection = errors.New("cannot inspect non-struct fields")
```

Comment:
This reads as "cannot inspect (non-struct fields)" instead of "cannot inspect non-struct (fields)" if that makes sense… It sounds like the error is complaining that the field content is not a struct